File: backend/internal/entity/ingredient.go
package entity

import "gorm.io/gorm"

type IngredientCategory struct {
	gorm.Model
	RestaurantID uint   `json:"restaurant_id" gorm:"not null;index;uniqueIndex:idx_ingredient_category_restaurant_name,priority:1"`
	Name         string `json:"name" gorm:"not null;size:120;uniqueIndex:idx_ingredient_category_restaurant_name,priority:2"`
	DisplayOrder int    `json:"display_order" gorm:"default:0"`
	IsActive     bool   `json:"is_active" gorm:"default:true"`

	Restaurant *Restaurant `json:"restaurant,omitempty" gorm:"foreignKey:RestaurantID"`
}

// IngredientUnitOption is one entry in a unit picker: a unit this ingredient
// accepts, and how many of its own stock units one of them makes.
type IngredientUnitOption struct {
	Unit         string  `json:"unit"`
	StockPerUnit float64 `json:"stock_per_unit"`
}

type Ingredient struct {
	gorm.Model
	RestaurantID uint    `json:"restaurant_id" gorm:"not null;index"`
	Name         string  `json:"name" gorm:"not null;size:160"`
	SKU          string  `json:"sku" gorm:"size:80;index"`
	CategoryID   *uint   `json:"category_id" gorm:"index"`
	ImageURL     string  `json:"image_url" gorm:"size:2048"`
	Unit         string  `json:"unit" gorm:"not null;size:40"`
	Stock        float64 `json:"stock" gorm:"type:numeric(18,4);not null;default:0;check:ingredient_stock_nonnegative,stock >= 0"`
	MinStock     float64 `json:"min_stock" gorm:"type:numeric(18,4);not null;default:0;check:ingredient_min_stock_nonnegative,min_stock >= 0"`
	// MaxStock is the most this shelf has been proved to hold: it rises to meet
	// any stock level above it and never falls. It is the denominator the apps
	// draw the level bar against — without it the bar's right-hand end means
	// nothing, which is what it meant before this column existed.
	MaxStock float64 `json:"max_stock" gorm:"type:numeric(18,4);not null;default:0;check:ingredient_max_stock_nonnegative,max_stock >= 0"`
	// MinPercent is the reorder level expressed as a share of MaxStock, and is
	// the source of truth when it is set: MinStock is then recomputed from it
	// every time MaxStock moves, so a shop that starts buying in larger loads
	// gets warned proportionally rather than at a quantity that is now too low.
	// Zero means nobody chose a percentage and MinStock stands on its own.
	MinPercent float64 `json:"min_percent" gorm:"type:numeric(7,4);not null;default:0;check:ingredient_min_percent_range,min_percent >= 0 AND min_percent <= 100"`
	CostPerUnit  float64 `json:"cost_per_unit" gorm:"type:numeric(14,4);not null;default:0;check:ingredient_cost_nonnegative,cost_per_unit >= 0"`
	YieldPercent float64 `json:"yield_percent" gorm:"type:numeric(7,4);not null;default:100;check:ingredient_yield_range,yield_percent > 0 AND yield_percent <= 100"`
	StorageType  string  `json:"storage_type" gorm:"size:40;default:room_temp"`

	Restaurant *Restaurant         `json:"restaurant,omitempty" gorm:"foreignKey:RestaurantID"`
	Category   *IngredientCategory `json:"category,omitempty" gorm:"foreignKey:CategoryID"`

	// DaysLeft and DailyUse are computed at read time, not stored: how long the
	// current stock lasts at the rate this ingredient was actually consumed over
	// the usage window. They are nil when nothing was consumed in the window —
	// there is no rate to divide by, and a 0 would read as "runs out today" when
	// the truth is "we have no idea yet". A client must render the nil case as
	// "no usage data", never as an empty bar.
	DaysLeft *float64 `json:"days_left,omitempty" gorm:"-"`
	DailyUse *float64 `json:"daily_use,omitempty" gorm:"-"`

	// UnitFamily is every unit a quantity may be entered in for this ingredient,
	// computed at read time from Unit. The client uses it to build the unit
	// picker, so the list of what converts to what lives in one place on the
	// server instead of being mirrored - and drifting - in each app.
	UnitFamily []IngredientUnitOption `json:"unit_family,omitempty" gorm:"-"`
}

type IngredientTransaction struct {
	gorm.Model
	RestaurantID uint    `json:"restaurant_id" gorm:"not null;index"`
	IngredientID uint    `json:"ingredient_id" gorm:"not null;index"`
	Type         string  `json:"type" gorm:"not null;size:16;check:ingredient_transaction_type,type IN ('in','out','adjust')"` // "in", "out", "adjust"
	Quantity     float64 `json:"quantity" gorm:"type:numeric(18,4);not null;check:ingredient_transaction_quantity_positive,quantity > 0"`
	// Amount is the money actually paid for a restock. Only "in" movements carry
	// it, and a positive value makes the stock-in write a matching Expense row —
	// that is the only way cash-out reaches the ledger without manual entry.
	Amount      float64 `json:"amount" gorm:"type:numeric(14,2);not null;default:0;check:ingredient_transaction_amount_nonnegative,amount >= 0"`
	Note        string  `json:"note" gorm:"size:500"`
	CreatedByID uint    `json:"created_by_id" gorm:"not null;index"`

	Ingredient *Ingredient `json:"ingredient,omitempty" gorm:"foreignKey:IngredientID"`
	CreatedBy  *User       `json:"-" gorm:"foreignKey:CreatedByID"`
}

File: frontend/src/types/ingredient.ts
export interface IngredientCategory {
  ID: number;
  restaurant_id: number;
  name: string;
  display_order: number;
  is_active: boolean;
  CreatedAt?: string;
  UpdatedAt?: string;
}

export interface Ingredient {
  ID: number;
  restaurant_id: number;
  name: string;
  // sku/image_url/yield_percent are no longer collected in the inventory form,
  // but the API still returns them and the menu page reads yield_percent for
  // per-dish cost — so the response type keeps them.
  sku?: string;
  category_id?: number | null;
  image_url?: string;
  unit: string;
  /** Every unit a quantity may be entered in for this ingredient, computed by
   *  the server from `unit`. Drives the unit picker; the list is authoritative,
   *  so the client never keeps its own conversion table. */
  unit_family?: IngredientUnitOption[];
  stock: number;
  min_stock: number;
  /**
   * The most this shelf has been proved to hold. It rises to meet any stock
   * level above it and never falls, so it is a real observed number rather than
   * a target anyone typed — which is what makes it safe to divide by. Zero
   * means nothing has been observed yet: draw no bar rather than invent one.
   */
  max_stock?: number;
  /**
   * The reorder level as a share of max_stock. When it is above zero the server
   * recomputes min_stock from it every time max_stock moves, so a shop that
   * starts buying in larger loads keeps being warned at the same proportion.
   */
  min_percent?: number;
  cost_per_unit: number;
  yield_percent?: number;
  storage_type?: string;
  /**
   * How many days the current stock lasts at the rate this ingredient was
   * actually consumed over the last 30 days. Computed at read time, absent when
   * nothing was consumed in the window — there is no rate to divide by, and a 0
   * would read as "runs out today". Render the absent case as "no usage data",
   * never as an empty bar.
   */
  days_left?: number;
  /** The daily rate days_left was derived from, for showing the working. */
  daily_use?: number;
  category?: IngredientCategory;
  CreatedAt?: string;
  UpdatedAt?: string;
}

export type TransactionType = "in" | "out" | "adjust";

export interface IngredientTransaction {
  ID: number;
  restaurant_id: number;
  ingredient_id: number;
  type: TransactionType;
  quantity: number;
  /** What a restock cost. Only "in" carries money; "out" and "adjust" are always 0. */
  amount: number;
  note: string;
  created_by_id: number;
  // The log stores ids only — the API joins these names on so a whole-inventory
  // read is readable. They are empty when the row points at a deleted record.
  ingredient_name: string;
  ingredient_unit: string;
  category_name: string;
  created_by_name: string;
  CreatedAt?: string;
}

export interface TransactionQuery {
  ingredient_id?: number;
  category_id?: number;
  type?: TransactionType | "";
  search?: string;
  /** Inclusive YYYY-MM-DD in the shop timezone. */
  from?: string;
  /** Inclusive YYYY-MM-DD — the API widens it to cover the whole day. */
  to?: string;
  page?: number;
  limit?: number;
}

export interface TransactionListResponse {
  transactions: IngredientTransaction[];
  total: number;
  page: number;
  limit: number;
}

export interface CSVExportResult {
  filename: string;
  /** Rows actually written. Fewer than `total` when the server capped the file. */
  rows: number;
  total: number;
  truncated: boolean;
}

export interface IngredientInput {
  name: string;
  category_id?: number;
  unit: string;
  stock: number;
  min_stock: number;
  /**
   * Omit it and the server leaves whatever percentage is stored alone; send 0
   * to clear it and keep min_stock as an absolute quantity. A client that never
   * sends it can therefore not erase a percentage set from somewhere else.
   */
  min_percent?: number;
  cost_per_unit: number;
  storage_type?: string;
}

/** One entry in a unit picker: a unit this ingredient accepts, and how many of
 *  its own stock units one of them makes (1 กิโลกรัม = 1000 against a gram shelf). */
export interface IngredientUnitOption {
  unit: string;
  stock_per_unit: number;
}

export interface AdjustStockInput {
  type: "in" | "out" | "adjust";
  quantity: number;
  /** Unit the quantity was typed in. Omitted means the ingredient's own unit.
   *  Any other unit of the same family is converted by the server. */
  unit?: string;
  note?: string;
  /** What the restock cost. Stock-in only; a positive value writes an expense entry. */
  amount?: number;
}
